internal/pkg/gateway/commit/notifier_test.go

```diff
@@ -0,0 +1,299 @@
+/*
+Copyright 2021 IBM All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package commit_test
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/golang/protobuf/proto"
+	commonProto "github.com/hyperledger/fabric-protos-go/common"
+	peerProto "github.com/hyperledger/fabric-protos-go/peer"
+	deliverMock "github.com/hyperledger/fabric/common/deliver/mock"
+	"github.com/hyperledger/fabric/internal/pkg/gateway/commit"
+	commitMocks "github.com/hyperledger/fabric/internal/pkg/gateway/commit/mocks"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestNotifier(t *testing.T) {
+	type Transaction struct {
+		ID     string
+		Status peerProto.TxValidationCode
+	}
+
+	NewPayload := func(transactionID string) *commonProto.Payload {
+		channelHeader := &commonProto.ChannelHeader{
+			Type: int32(commonProto.HeaderType_ENDORSER_TRANSACTION),
+			TxId: transactionID,
+		}
+		channelHeaderBytes, err := proto.Marshal(channelHeader)
+		assert.NoError(t, err)
+
+		return &commonProto.Payload{
+			Header: &commonProto.Header{
+				ChannelHeader: channelHeaderBytes,
+			},
+		}
+	}
+
+	NewTransactionEnvelope := func(transaction *Transaction) *commonProto.Envelope {
+		payload := NewPayload(transaction.ID)
+		payloadBytes, err := proto.Marshal(payload)
+		assert.NoError(t, err)
+
+		return &commonProto.Envelope{
+			Payload: payloadBytes,
+		}
+	}
+
+	NewBlock := func(t *testing.T, transactions ...*Transaction) *commonProto.Block {
+		var validationCodes []byte
+		var envelopes [][]byte
+
+		for _, transaction := range transactions {
+			validationCodes = append(validationCodes, byte(transaction.Status))
+
+			envelope := NewTransactionEnvelope(transaction)
+			envelopeBytes, err := proto.Marshal(envelope)
+			assert.NoError(t, err)
+			envelopes = append(envelopes, envelopeBytes)
+		}
+
+		metadata := make([][]byte, len(commonProto.BlockMetadataIndex_name))
+		metadata[int(commonProto.BlockMetadataIndex_TRANSACTIONS_FILTER)] = validationCodes
+
+		return &commonProto.Block{
+			Header: &commonProto.BlockHeader{
+				Number: 1,
+			},
+			Metadata: &commonProto.BlockMetadata{
+				Metadata: metadata,
+			},
+			Data: &commonProto.BlockData{
+				Data: envelopes,
+			},
+		}
+	}
+
+	NewMockBlockReader := func(blocks ...*commonProto.Block) *commitMocks.BlockReader {
+		blockReader := &commitMocks.BlockReader{}
+
+		iterator := &deliverMock.BlockIterator{}
+		blockReader.IteratorReturns(iterator, nil)
+
+		for i, block := range blocks {
+			iterator.NextReturnsOnCall(i, block, commonProto.Status_SUCCESS)
+		}
+
+		return blockReader
+	}
+
+	NewTestNotifier := func(blocks ...*commonProto.Block) *commit.Notifier {
+		blockReader := NewMockBlockReader(blocks...)
+		return commit.NewNotifier(blockReader)
+	}
+
+	t.Run("Notify returns error if block iterator cannot be obtained", func(t *testing.T) {
```

Comment:
There's nothing strictly wrong with the pattern you're following here, embedding all of the setup functions using closures which capture the *testing.T in order to not have to pass it around.  That being said, it would also have been perfectly acceptable (and perhaps more typical) to simply make the closures vanilla functions which take a `*testing.T` as their first parameter since presumably, they will be useful in other tests.  But, no need to change it now.

Additionally, it seems possible that  each of these test cases could be consolidated into a table test like: https://github.com/hyperledger/fabric/blob/5bad6aec88e576a377a65192a4eaf69e8307f5f7/core/chaincode/chaincode_support_test.go#L1218-L1290

I'm not sure how perfectly these would transition into a table test, but there seems to be quite a bit of shared setup, followed by a set of assertions over the same results.  Not a requirement, but you might want to see if it makes sense.

Reply:
Good spot. The direct use of `t` in the closure is a bug in my test code and would fail at runtime if any of the assertions in the helper functions tripped up. I'll fix this

Reply:
I think there is too much variation in the set up and assertions across these unit tests for table testing to be useful here, but I'll look again